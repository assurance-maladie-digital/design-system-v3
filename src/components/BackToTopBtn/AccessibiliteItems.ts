import { ExpertiseLevelEnum } from './constants/ExpertiseLevelEnum'

export const AccessibiliteItemsIndeterminate = [
	
	{
		title: 'Catégorie 10 : Présentation de l\'information',
		subtitle: '10.8 Pour chaque page web, les contenus cachés ont-ils vocation à être ignorés par les technologies d’assistance',
		items: [
			{

				precision: '10.8.1 Le contenu caché a vocation à être ignoré par les technologies d’assistance , il n’a pas vocation à être ignoré par les technologies d’assistance et est rendu restituable par les technologies d’assistance suite à une action de l’utilisateur réalisable au clavier ou par tout dispositif de pointage sur un élément précédent le contenu caché ou suite à un repositionnement du focus dessus',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#10.8.1',
				solution: [{
					info1: '1. Retrouver les contenus cachés (éléments pourvus de l’attribut hidden ou de l’attribut WAI-ARIA aria-hidden, ou bien d’une classe ou d’un ensemble de styles CSS susceptibles de masquer le contenu)',
					info2: '2. Pour chaque contenu caché, vérifier que :'
						+ 'Soit le contenu caché a vocation à être ignoré par les technologies d’assistance (un élément statistique de visites par exemple) ;'
						+ 'Soit le contenu caché n’a pas vocation à être ignoré par les technologies d’assistance, et dans ce cas il est rendu restituable par les technologies d’assistance au moyen :'
						+ 'Soit d’une action de l’utilisateur réalisable au clavier ou par tout dispositif de pointage sur un élément précédent le contenu caché ;'
						+ 'Soit d’une fonction de programmation qui repositionne le focus sur le contenu',
					info3: '3. Si c’est le cas pour chaque contenu caché, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DEV,

			},

		],
	},
]

export const AccessibiliteItemsValidated = [

	{
		title: 'Catégorie 3 : Couleurs',
		subtitle: '3.2 Dans chaque page web, le contraste entre la couleur du texte et la couleur de son arrière-plan est-il suffisamment élevé?',
		items: [
			{

				precision: '3.2.1 Dans chaque page web, le texte et le texte en image sans effet de graisse d’une taille restituée inférieure à 24px vérifient-ils une de ces conditions? Le rapport de contraste entre le texte et son arrière-plan est de 4.5:1, au moins; Un mécanisme permet à l’utilisateur d’afficher le texte avec un rapport de contraste de 4.5:1, au moins.',
				link: 'https://accessibilite.numerique.gouv.fr/methode/criteres-et-tests/#3.2.1',
				solution: [{
					info1: '1. Retrouver dans le document les textes et les textes en image sans effet de graisse d’une taille restituée inférieure à 24px qui pourraient poser des problèmes de contraste ;',
					info2: '2. Pour chacun de ces textes, vérifier que : Soit le rapport de contraste entre le texte et son arrière-plan est de 4.5:1, au moins; Soit un mécanisme permet à l’utilisateur d’afficher le texte avec un rapport de contraste de 4.5:1, au moins.',
					info3: '3. Si c’est le cas pour chaque texte, le test est validé',
				}],
				expertise: ExpertiseLevelEnum.DESIGN,

			},

		],
	},

]
