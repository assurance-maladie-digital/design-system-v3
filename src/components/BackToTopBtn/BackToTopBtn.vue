<script setup lang="ts">
	import { mdiArrowUp } from '@mdi/js'
	import { computed, onMounted, onUnmounted, ref } from 'vue'

	import useCustomizableOptions, { type CustomizableOptions } from '@/composables/useCustomizableOptions'
	import { convertToUnit } from '@/utils/convertToUnit'
	import { useDisplay } from 'vuetify'

	import { locales } from './locales'
	import { config } from './config'

	import { type VBtn } from 'vuetify/components'

	type VBtnProps = InstanceType<typeof VBtn>['$props']

	export interface Props {
		threshold?: number
		nudgeRight?: string | number
		nudgeBottom?: string | number
		target?: string
	}

	const props = withDefaults(defineProps<Props & CustomizableOptions>(), {
		threshold: 120,
		nudgeRight: '16px',
		nudgeBottom: '16px',
		target: undefined,
	})

	const options = useCustomizableOptions(config, props)

	const showBtn = ref(false)

	const targetSelector = computed(() => {
		return props.target ? `#${props.target}` : null
	})

	const isMobile = computed(() => useDisplay().smAndDown.value)

	const btnStyle = computed(() => {
		const right = convertToUnit(props.nudgeRight) || '0'
		const bottom = convertToUnit(props.nudgeBottom) || '0'

		return {
			bottom,
			marginRight: right,
		}
	})

	const minWidth = computed(() => (isMobile.value ? '36px' : undefined))

	const labelClasses = computed(() => ({
		'd-sr-only': isMobile.value,
	}))

	// Methods
	const onScroll = (e: Event) => {
		const target = e.currentTarget as HTMLElement | Window

		if (target instanceof Window) {
			showBtn.value = window.scrollY > props.threshold
		}
		else {
			showBtn.value = (target as HTMLElement).scrollTop > props.threshold
		}
	}

	const scrollToTop = () => {
		if (!props.target) {
			window.scrollTo(0, 0)
		}
		else {
			const target = document.getElementById(props.target) || window
			target.scrollTo(0, 0)
		}
	}

	// Handling scroll events
	const target = computed(() => props.target ? document.getElementById(props.target) : window)
	onMounted(() => {
		target.value?.addEventListener('scroll', onScroll)
	})
	onUnmounted(() => {
		target.value?.removeEventListener('scroll', onScroll)
	})

</script>

<template>
	<VFadeTransition>
		<VBtn
			v-show="showBtn"
			v-scroll:[targetSelector]="onScroll"
			v-bind="({
				...options.btn,
				...$attrs,
			} as VBtnProps)"
			:style="btnStyle"
			:min-width="minWidth"
			class="vd-back-to-top-btn"
			@click="scrollToTop"
		>
			<span :class="labelClasses">
				<slot>
					{{ locales.label }}
				</slot>
			</span>

			<slot name="icon">
				<VIcon v-bind="options.icon">
					{{ mdiArrowUp }}
				</VIcon>
			</slot>
		</VBtn>
	</VFadeTransition>
</template>

<style lang="scss" scoped>
.vd-back-to-top-btn {
	position: sticky;
	z-index: 999;
	opacity: 1;
	float: right;
}
.v-btn--variant-outlined {
	background: white;
}
</style>
