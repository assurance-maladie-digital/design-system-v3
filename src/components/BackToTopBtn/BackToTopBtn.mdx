import {Controls, Canvas, Meta, Source} from '@storybook/blocks';

import * as BackToTopBtnStories from './BackToTopBtn.stories';

<Meta of={BackToTopBtnStories} />

# BackToTopBtn

Le composant `BackToTopBtn` est utilisé pour afficher un bouton permettant à l’utilisateur de remonter en haut de la page.


Le boutton apparaît lorsque l'utilisateur scroll en dessous d'une certaine hauteur de la page. Cette hauteur est définie par la prop `threshold`, elle est de 120px par defaut.


<Canvas of={BackToTopBtnStories.Default}  />

## API

<Controls of={BackToTopBtnStories.Default} />

## Context d'utilisation

Ce composant peut être utilisé à la racine de la page ou bien dans un composant avant son propre context de scroll. Dans ce dernier cas, il conviens de passer la prop `target` un id unique de l'élément scrollable.

### Exemple d'utilisation

<Source dark code={`
<div
    id="test-container"
    style="height: 600px; overflow: scroll; position: relative; border: 1px solid black; scroll-behavior: smooth;"
>
    <p>lorem ipsum</p>
    <div style="height: 3000px" />
    <BackToTopBtn target='test-container'>
        Retour en haut de la page
    </BackToTopBtn>
</div>
`} />

## Customisation

Ce composant utilise en interne les composants `VBtn` et `VIcon` de Vuetify. 
Il est possible de passer des props à ces composants pour les customiser en utilisant la props `vuetifyOptions`.

<Canvas of={BackToTopBtnStories.Customization}  />
