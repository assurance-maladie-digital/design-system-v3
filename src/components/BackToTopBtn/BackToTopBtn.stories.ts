import type { Meta, StoryObj } from '@storybook/vue3'

import BackToTopBtn from './BackToTopBtn.vue'

const meta = {
	title: 'Components/BackToTopBtn',
	component: BackToTopBtn,
	parameters: {
		layout: 'fullscreen',
	},
	argTypes: {
		threshold: {
			control: { type: 'range', max: 2000 },
		},
		nudgeRight: {
			control: { type: 'range' },
			default: 16,
		},
		nudgeBottom: {
			control: { type: 'range' },
			default: 16,
		},
		target: {
			control: { type: 'text' },
		},
		vuetifyOptions: {
			control: { type: 'object' },
			description: 'Customisez les options des composants internes de Vuetify',
			table: {
				category: 'props',
			},
		},
		default: {
			control: { type: 'text' },
		},
		icon: {
			control: { type: 'text' },
		},
	},
	args: {
		threshold: 120,
		nudgeRight: 16,
		nudgeBottom: 16,
		target: 'testcontainer',
		default: 'Retour en haut',
	},
} satisfies Meta<typeof BackToTopBtn>

export default meta

type Story = StoryObj<typeof meta>

export const Default: Story = {
	render: (args) => {
		return {
			components: {
				BackToTopBtn,
			},
			setup() {
				return { args }
			},
			template: `
				<div
					id="testcontainer"
					style="height: 60vh; overflow: scroll; position: relative; border: 1px solid black; scroll-behavior: smooth; background: #fff"
				>
					<p>Scroll down</p>
					<div style="height: 360vh" />
					<BackToTopBtn
						v-bind="args"
					>
						{{args.default}}
					</BackToTopBtn>
				</div>
			`,
		}
	},
	play: async ({ canvasElement }) => {
		await new Promise((resolve: (v: unknown) => void) => setTimeout(resolve, 2000))
		const container = canvasElement.querySelector('#testcontainer')
		container?.scrollTo(0, 1000)
	},
}

export const Customization: Story = {
	args: {
		target: 'btn-customization',
		vuetifyOptions: {
			btn: {
				variant: 'tonal',
				color: 'red',
				rounded: false,
			},
			icon: {
				color: 'red',
			},
		},
	},
	render: (args) => {
		return {
			components: {
				BackToTopBtn,
			},
			setup() {
				return { args }
			},
			template: `
				<div
					id="btn-customization"
					style="height: 60vh; overflow: scroll; position: relative; border: 1px solid black; scroll-behavior: smooth; background: #fff"
				>
					<p>Scroll down</p>
					<div style="height: 360vh" />
					<BackToTopBtn
						v-bind="args"
					>
						{{args.default}}
					</BackToTopBtn>
				</div>
			`,
		}
	},
	play: async ({ canvasElement }) => {
		await new Promise((resolve: (v: unknown) => void) => setTimeout(resolve, 2000))
		const container = canvasElement.querySelector('#btn-customization')
		container?.scrollTo(0, 1000)
	},
}
