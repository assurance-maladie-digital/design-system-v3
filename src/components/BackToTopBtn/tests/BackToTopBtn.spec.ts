import { describe, it, expect, vi, afterEach } from 'vitest'
import { mount } from '@vue/test-utils'

import BackToTopBtn from '../'
import { vuetify } from '@tests/unit/setup'

describe('BackToTopBtn', () => {
	afterEach(() => {
		vi.restoreAllMocks()
		document.body.innerHTML = ''
	})

	it('renders correctly', () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('renders correctly when nudgeBottom and nudgeRight are set to invalid values', () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
			props: {
				nudgeBottom: '',
				nudgeRight: '',
			},
		})

		expect(wrapper.html()).toMatchSnapshot()
	})

	it('shows the button when the user scrolls down', async () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
			attachTo: document.body,
		})

		const buttonElement = wrapper.find('.vd-back-to-top-btn')
		expect(buttonElement.isVisible()).toBeFalsy()

		vi.spyOn(window, 'scrollY', 'get').mockReturnValue(500)
		window.dispatchEvent(new CustomEvent('scroll'))
		await wrapper.vm.$nextTick()

		expect(buttonElement.isVisible()).toBeTruthy()
	})

	it('shows the button when the user scrolls down and the target is a custom element', async () => {
		document.body.innerHTML = '<div id="test"></div>'
		const divContent = document.getElementById('test') as HTMLDivElement

		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
			attachTo: divContent,
			props: {
				target: 'test',
			},
		})

		const buttonElement = wrapper.find('.vd-back-to-top-btn')
		expect(buttonElement.isVisible()).toBeFalsy()

		vi.spyOn(divContent, 'scrollTop', 'get').mockReturnValue(500)
		divContent.dispatchEvent(new CustomEvent('scroll'))

		await wrapper.vm.$nextTick()
		expect(buttonElement.isVisible()).toBeTruthy()
	})

	it('do not show the button when the user scrolls up', async () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
			attachTo: document.body,
		})

		const buttonElement = wrapper.find('.vd-back-to-top-btn')
		expect(buttonElement.isVisible()).toBeFalsy()

		vi.spyOn(window, 'scrollY', 'get').mockReturnValue(500)
		window.dispatchEvent(new CustomEvent('scroll'))
		await wrapper.vm.$nextTick()
		expect(buttonElement.isVisible()).toBeTruthy()

		vi.spyOn(window, 'scrollY', 'get').mockReturnValue(100)
		window.dispatchEvent(new CustomEvent('scroll'))
		await wrapper.vm.$nextTick()

		expect(buttonElement.isVisible()).toBeFalsy()
	})

	it('scrolls to the top when the button is clicked', async () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
		})

		const scrollToSpy = vi
			.spyOn(window, 'scrollTo')
			.mockImplementation(() => {})

		await wrapper.find('.v-btn').trigger('click')

		expect(scrollToSpy).toHaveBeenCalledWith(0, 0)
	})

	it('scrolls to the top when the button is clicked and the target is a custom element', async () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				target: 'test',
			},
		})

		const scrollToSpy = vi.fn()

		vi.spyOn(document, 'getElementById').mockReturnValue({
			scrollTo: scrollToSpy,
		} as unknown as HTMLElement)

		await wrapper.find('.v-btn').trigger('click')

		expect(scrollToSpy).toHaveBeenCalledWith(0, 0)
	})

	it('scrolls to the top when the button is clicked and the target is a custom element that does not exist', async () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
			propsData: {
				target: 'test',
			},
		})

		const scrollToSpy = vi
			.spyOn(window, 'scrollTo')
			.mockImplementation(() => {})

		vi.spyOn(document, 'getElementById').mockReturnValue(null)

		await wrapper.find('.v-btn').trigger('click')

		expect(scrollToSpy).toHaveBeenCalledWith(0, 0)
	})

	it('remove the scroll event listener when the component is destroyed', async () => {
		const wrapper = mount(BackToTopBtn, {
			global: {
				plugins: [vuetify],
			},
		})

		const removeEventListenerSpy = vi.spyOn(window, 'removeEventListener')

		wrapper.unmount()

		expect(removeEventListenerSpy).toHaveBeenCalledWith('scroll', expect.any(Function))
	})
})
