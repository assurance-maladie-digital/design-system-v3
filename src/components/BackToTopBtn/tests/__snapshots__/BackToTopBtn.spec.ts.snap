// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BackToTopBtn > renders correctly 1`] = `
"<transition-stub data-v-8ea81d33="" name="fade-transition" appear="false" persisted="false" css="true"><button data-v-8ea81d33="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-outlined text-wrap px-0 px-md-4 vd-back-to-top-btn" style="bottom: 16px; margin-right: 16px; display: none;"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator=""><span data-v-8ea81d33="" class="">Retour en haut</span><i data-v-8ea81d33="" class="M13,20H11V8L5.5,13.5L4.08,12.08L12,4.16L19.92,12.08L18.5,13.5L13,8V20Z mdi v-icon notranslate v-theme--light v-icon--size-small text-primary ml-0 ml-md-1" aria-hidden="true"></i></span>
    <!---->
    <!---->
  </button></transition-stub>"
`;

exports[`BackToTopBtn > renders correctly when nudgeBottom and nudgeRight are set to invalid values 1`] = `
"<transition-stub data-v-8ea81d33="" name="fade-transition" appear="false" persisted="false" css="true"><button data-v-8ea81d33="" type="button" class="v-btn v-theme--light text-primary v-btn--density-default v-btn--size-default v-btn--variant-outlined text-wrap px-0 px-md-4 vd-back-to-top-btn" style="bottom: 0px; margin-right: 0px; display: none;"><span class="v-btn__overlay"></span><span class="v-btn__underlay"></span>
    <!----><span class="v-btn__content" data-no-activator=""><span data-v-8ea81d33="" class="">Retour en haut</span><i data-v-8ea81d33="" class="M13,20H11V8L5.5,13.5L4.08,12.08L12,4.16L19.92,12.08L18.5,13.5L13,8V20Z mdi v-icon notranslate v-theme--light v-icon--size-small text-primary ml-0 ml-md-1" aria-hidden="true"></i></span>
    <!---->
    <!---->
  </button></transition-stub>"
`;
